diff --git a/cluster_run b/cluster_run
index f06dbfa..d54eab7 100755
--- a/cluster_run
+++ b/cluster_run
@@ -9,6 +9,7 @@ base_projector_port = 10000
 base_xdcr_port = 13000
 base_indexer_port = 9100
 base_fts_port = 9200
+base_eventing_port = 25000
 
 LOGLEVELS = ["debug", "info", "warn", "error", "critical"]
 
@@ -158,7 +159,8 @@ def erlang_args_for_node(i):
         "indexer_stcatchup_port", str(base_indexer_port + i * 6 + 4),
         "indexer_stmaint_port", str(base_indexer_port + i * 6 + 5),
         "fts_http_port", str(base_fts_port + i),
-        "fts_ssl_port", str(10000 + base_fts_port + i)
+        "fts_ssl_port", str(10000 + base_fts_port + i),
+        "eventing_admin_port", str(base_eventing_port + i)
         ] + cluster_extra_args
 
     return args
