diff --git a/cluster_connect b/cluster_connect
index d98bbea..f2fbfaa 100755
--- a/cluster_connect
+++ b/cluster_connect
@@ -3,8 +3,8 @@ import os, sys, getopt, urllib2, urllib, json
 
 from urllib2 import HTTPError
 
-valid_bucket_types = ["ephemeral", "membase", "memcached"]
-valid_service_types = {"kv", "n1ql", "index", "fts", "example"}
+valid_bucket_types = ["membase", "memcached"]
+valid_service_types = {"kv", "n1ql", "index", "fts", "example", "eventing"}
 
 def usage():
     print "usage: \n\
@@ -12,8 +12,8 @@ def usage():
             -T <services to run. kv if unspecified> (eg: n0:kv,n1:index+n1ql+fts)\n\
             -s <memory size (min 256) default: 256\n\
             -I <index memory size> default: 256\n\
-            -t <bucket type> (ephemeral, membase, memcached) default: membase\n\
-            -r <num replicas> (max 3) default: 1 (Only for ephemeral or membase buckets!)\n\
+            -t <bucket type> (membase, memcached) default: membase\n\
+            -r <num replicas> (max 3) default: 1 (Only for membase buckets!)\n\
             -i (don't index replicas) default: replica index enabled\n\
             -S <start index> default: 0"
 
@@ -36,12 +36,12 @@ def main():
         sys.exit()
     nodes = 0
     buckettype = "membase"
-    memsize = 256
+    memsize = 512
     indexmemsize = 256
-    replicas = 1
-    replica_index = True
+    replicas = 0
+    replica_index = False
     start_index = 0
-    deploy = ['kv']
+    deploy = ['eventing', 'index', 'n1ql', 'fts']
     do_rebalance = True
 
     data_base_path = os.getcwd() + "/data"
@@ -110,7 +110,23 @@ def main():
            "&authType=sasl" +
            "&saslPassword=" +
            "&bucketType=" + buckettype +
-           "&ramQuotaMB=" + str(memsize) +
+           "&ramQuotaMB=" + str(100) +
+           "&replicaNumber=" + str(replicas) +
+           "&replicaIndex=" + bool_request_value(replica_index)).read()
+    o.open("http://127.0.0.1:{0}/pools/default/buckets".format(base_port),
+           "name=eventing" +
+           "&authType=sasl" +
+           "&saslPassword=" +
+           "&bucketType=" + buckettype +
+           "&ramQuotaMB=" + str(100) +
+           "&replicaNumber=" + str(replicas) +
+           "&replicaIndex=" + bool_request_value(replica_index)).read()
+    o.open("http://127.0.0.1:{0}/pools/default/buckets".format(base_port),
+           "name=credit_score" +
+           "&authType=sasl" +
+           "&saslPassword=" +
+           "&bucketType=" + buckettype +
+           "&ramQuotaMB=" + str(100) +
            "&replicaNumber=" + str(replicas) +
            "&replicaIndex=" + bool_request_value(replica_index)).read()
     o.open("http://127.0.0.1:{0}/settings/web".format(base_port),
