diff --git a/secondary/dcp/conn_pool.go b/secondary/dcp/conn_pool.go
index 44e0e22c..a17e3e72 100644
--- a/secondary/dcp/conn_pool.go
+++ b/secondary/dcp/conn_pool.go
@@ -204,7 +204,7 @@ func (cp *connectionPool) StartTapFeed(args *memcached.TapArguments) (*memcached
 const DEFAULT_WINDOW_SIZE = uint32(20 * 1024 * 1024) // 20 Mb
 
 func (cp *connectionPool) StartDcpFeed(
-	name DcpFeedName, sequence uint32,
+	name DcpFeedName, sequence, flags uint32,
 	outch chan *memcached.DcpEvent,
 	opaque uint16,
 	config map[string]interface{}) (*memcached.DcpFeed, error) {
@@ -223,7 +223,9 @@ func (cp *connectionPool) StartDcpFeed(
 
 	dcpf, err := memcached.NewDcpFeed(mc, string(name), outch, opaque, config)
 	if err == nil {
-		err = dcpf.DcpOpen(string(name), sequence, DEFAULT_WINDOW_SIZE, opaque)
+		err = dcpf.DcpOpen(
+			string(name), sequence, flags, DEFAULT_WINDOW_SIZE, opaque,
+		)
 		if err == nil {
 			return dcpf, err
 		}
