diff --git a/priv/public/js/app-misc.js b/priv/public/js/app-misc.js
index 5968dfa..ed32752 100644
--- a/priv/public/js/app-misc.js
+++ b/priv/public/js/app-misc.js
@@ -441,6 +441,7 @@ _.extend(ViewHelpers, {
         case 'moxi': return 'Moxi';
         case 'index': return 'Index';
         case 'fts': return 'Full Text';
+        case 'eventing': return 'Eventing';
       }
     }).sort();
   },
diff --git a/priv/public/ui/app/components/directives/mn_services/mn_services.html b/priv/public/ui/app/components/directives/mn_services/mn_services.html
index 54a7733..4b7aa8e 100644
--- a/priv/public/ui/app/components/directives/mn_services/mn_services.html
+++ b/priv/public/ui/app/components/directives/mn_services/mn_services.html
@@ -46,3 +46,13 @@
 <label for="for-fts-service">
   Full Text
 </label>
+<input
+    type="checkbox"
+    value="eventing"
+    id="eventing_service"
+    ng-change="onChange(config.services.model.eventing, 'eventing')"
+    ng-model="config.services.model.eventing"
+    ng-disabled="config.services.disabled.eventing">
+<label for="eventing_service">
+    Eventing
+</label>
diff --git a/priv/public/ui/app/components/mn_filters.js b/priv/public/ui/app/components/mn_filters.js
index 38f2f82..1b7201a 100644
--- a/priv/public/ui/app/components/mn_filters.js
+++ b/priv/public/ui/app/components/mn_filters.js
@@ -491,6 +491,7 @@
         case 'n1ql': return 'Query';
         case 'index': return 'Index';
         case 'fts': return 'Full Text';
+        case 'eventing': return 'Eventing';
       }
     }
   }
