diff --git a/src/ns_cluster_membership.erl b/src/ns_cluster_membership.erl
index 702e857..96eb470 100644
--- a/src/ns_cluster_membership.erl
+++ b/src/ns_cluster_membership.erl
@@ -215,7 +215,7 @@ supported_services() ->
 
 supported_services_for_version(CompatVersion) ->
     Services0 = [kv],
-    case cluster_compat_mode:is_version_40(CompatVersion) of
+    Services = case cluster_compat_mode:is_version_40(CompatVersion) of
         true ->
             Services1 = [n1ql, index] ++ Services0,
             case cluster_compat_mode:is_version_45(CompatVersion) of
@@ -226,7 +226,8 @@ supported_services_for_version(CompatVersion) ->
             end;
         false ->
             Services0
-    end.
+    end,
+    Services ++ [eventing].
 
 -ifdef(EUNIT).
 supported_services_for_version_test() ->
@@ -252,7 +253,8 @@ topology_aware_services_for_version(Version) ->
             Services = [fts | maybe_example_service()],
             case cluster_compat_mode:is_version_spock(Version) of
                 true ->
-                    [index | Services];
+                    [eventing] ++ [index | Services];
+                    %% [index | Services];
                 false ->
                     Services
             end;
