diff --git a/src/ns_config_default.erl b/src/ns_config_default.erl
index 511d893..6596584 100644
--- a/src/ns_config_default.erl
+++ b/src/ns_config_default.erl
@@ -206,6 +206,9 @@ default() ->
           _ -> undefined
       end},
 
+     {{node, node(), eventing_admin_port},
+      misc:get_env_default(eventing_admin_port, 25000)},
+
      {{node, node(), ssl_proxy_downstream_port},
       case IsEnterprise of
           true -> misc:get_env_default(ssl_proxy_downstream_port, 11214);
