diff --git a/src/ns_ports_setup.erl b/src/ns_ports_setup.erl
index 6632049..150e86f 100644
--- a/src/ns_ports_setup.erl
+++ b/src/ns_ports_setup.erl
@@ -275,6 +275,7 @@ do_dynamic_children(normal, Config) ->
      per_bucket_moxi_specs(Config),
      maybe_create_ssl_proxy_spec(Config),
      fts_spec(Config),
+     eventing_spec(Config),
      example_service_spec(Config)].
 
 expand_specs(Specs, Config) ->
@@ -630,6 +631,32 @@ fts_spec(Config) ->
             [Spec]
     end.
 
+eventing_spec(Config) ->
+    case ns_cluster_membership:should_run_service(Config, eventing, node()) of
+        false ->
+            [];
+        _ ->
+            Command = path_config:component_path(bin, "eventing"),
+            EventingAdminPort = ns_config:search(Config, {node, node(), eventing_admin_port}, 25000),
+            {ok, EventingDir} = ns_storage_conf:this_node_ixdir(),
+            LocalMemcachedPort = ns_config:search_node_prop(node(), Config, memcached, port),
+            NodeUUID = ns_config:search(Config, {node, node(), uuid}, false),
+            RestPort = misc:node_rest_port(Config, node()),
+            EventingDir2 = filename:join(EventingDir, "@eventing"),
+            EventingAdminArg = "-adminport=" ++ integer_to_list(EventingAdminPort),
+            EventingDirArg = "-dir=" ++ EventingDir2,
+            KVAddrArg = "-kvport=" ++ integer_to_list(LocalMemcachedPort),
+            RestArg = "-restport=" ++ integer_to_list(RestPort),
+            UUIDArg = "-uuid=" ++ NodeUUID,
+
+            Spec = {'eventing', Command,
+                    [EventingAdminArg, EventingDirArg, KVAddrArg, RestArg, UUIDArg],
+                    [via_goport, exit_status, stderr_to_stdout,
+                     {env, build_go_env_vars(Config, eventing)},
+                     {log, "eventing.log"}]},
+            [Spec]
+    end.
+
 example_service_spec(Config) ->
     CacheCmd = find_executable("cache-service"),
     NodeUUID = ns_config:search(Config, {node, node(), uuid}, false),
