diff --git a/plasma/plasma.go b/plasma/plasma.go
index 8cda566..a85e672 100644
--- a/plasma/plasma.go
+++ b/plasma/plasma.go
@@ -359,11 +359,11 @@ func New(cfg Config) (*Plasma, error) {
 		}
 	}
 
-	if s.shouldPersist {
-		go s.monitorMemUsage()
-	}
+	// if s.shouldPersist {
+	// 	go s.monitorMemUsage()
+	// }
 
-	go s.runtimeStats()
+	// go s.runtimeStats()
 	return s, err
 }
 
@@ -1030,7 +1030,7 @@ retry:
 	}
 
 refresh:
-	s.tryThrottleForResources(ctx)
+	// s.tryThrottleForResources(ctx)
 
 	if pg, err = s.ReadPage(pid, ctx.pgRdrFn, false, ctx); err != nil {
 		return nil, nil, err
