diff --git a/secondary/dcp/transport/client/transport.go b/secondary/dcp/transport/client/transport.go
index 84e078ed..0a456660 100644
--- a/secondary/dcp/transport/client/transport.go
+++ b/secondary/dcp/transport/client/transport.go
@@ -57,3 +57,11 @@ func transmitRequest(o io.Writer, req *transport.MCRequest) (int, error) {
 	}
 	return n, err
 }
+
+func transmitResponse(o io.Writer, res *transport.MCResponse) (int, error) {
+	if o == nil {
+		return 0, errNoConn
+	}
+	n, err := res.Transmit(o)
+	return n, err
+}
