diff --git a/secondary/dcp/upr.go b/secondary/dcp/upr.go
index 6e7fbe74..b6ccda3f 100644
--- a/secondary/dcp/upr.go
+++ b/secondary/dcp/upr.go
@@ -76,7 +76,9 @@ func (b *Bucket) GetFailoverLogs(
 		}
 
 		name := NewDcpFeedName(fmt.Sprintf("getfailoverlog-%s-%v", b.Name, time.Now().UnixNano()))
-		singleFeed, err := serverConn.StartDcpFeed(name, 0, nil, opaque, config)
+		flags := uint32(0x0)
+		singleFeed, err := serverConn.StartDcpFeed(
+			name, 0, flags, nil, opaque, config)
 		if err != nil {
 			return nil, err
 		}
@@ -120,11 +122,11 @@ type DcpFeed struct {
 // No data will be sent on the channel unless vbuckets streams
 // are requested.
 func (b *Bucket) StartDcpFeed(
-	name DcpFeedName, sequence uint32,
+	name DcpFeedName, sequence, flags uint32,
 	opaque uint16,
 	config map[string]interface{}) (*DcpFeed, error) {
 
-	return b.StartDcpFeedOver(name, sequence, nil, opaque, config)
+	return b.StartDcpFeedOver(name, sequence, flags, nil, opaque, config)
 }
 
 // StartDcpFeedOver creates and starts a new Dcp feed.
@@ -139,7 +141,7 @@ func (b *Bucket) StartDcpFeed(
 //      "numConnections", number of connections with DCP for local vbuckets.
 func (b *Bucket) StartDcpFeedOver(
 	name DcpFeedName,
-	sequence uint32,
+	sequence, flags uint32,
 	kvaddrs []string,
 	opaque uint16,
 	config map[string]interface{}) (*DcpFeed, error) {
@@ -159,7 +161,7 @@ func (b *Bucket) StartDcpFeedOver(
 	}
 	feed.numConnections = config["numConnections"].(int)
 	feed.C = feed.output
-	if feed.connectToNodes(kvaddrs, opaque, config) != nil {
+	if feed.connectToNodes(kvaddrs, opaque, flags, config) != nil {
 		return nil, ErrorInvalidBucket
 	}
 	go feed.genServer(feed.reqch, opaque)
@@ -283,7 +285,8 @@ loop:
 }
 
 func (feed *DcpFeed) connectToNodes(
-	kvaddrs []string, opaque uint16, config map[string]interface{}) error {
+	kvaddrs []string, opaque uint16, flags uint32,
+	config map[string]interface{}) error {
 
 	prefix := feed.logPrefix
 	kvcache := make(map[string]bool)
@@ -318,7 +321,7 @@ func (feed *DcpFeed) connectToNodes(
 		for i := 0; i < feed.numConnections; i++ {
 			feedname := DcpFeedName(fmt.Sprintf("%v/%d", name, i))
 			singleFeed, err := serverConn.StartDcpFeed(
-				feedname, feed.sequence, feed.output, opaque, config)
+				feedname, feed.sequence, flags, feed.output, opaque, config)
 			if err != nil {
 				for _, singleFeed := range nodeFeeds {
 					singleFeed.dcpFeed.Close()
